### ISSUE TITLE
Events being dropped on certain machines

### ISSUE THREAD

**Issue by hypnobunny** [sarl/sarl#713]
Hi SARL team,

While running a medium-scale multi-agent system, we appear to be hitting a limit on how many events we can send at any one time on certain machines. In our multi-agent scenario, we have approximately 12 agents in a given space, and a single 'Communicator' agent receiving XML data from an external source once per second. This Communicator transforms the XML data into a series of different granular events and emits to the default space. As such, each second, approximately 40-60 events are being emitted. 

Under these conditions, on one of our machines (Microsoft Surface Pro 4, 8GB RAM, i5 dual core processor, 256GB SSD) we begin receiving error messages stating:

```[SEVERE, 7:30:03pm, null] Rejected scheduled task: io.janusproject.kernel.bic.internaleventdispatching.AgentInternalEventsDispatcher$$Lambda$18/184144855@7f917118```

Here is an example screenshot showing the errors:

![errors](https://user-images.githubusercontent.com/3704446/29344568-1bf8239a-827c-11e7-96d0-7bfa658bca42.png)

Having done some elementary research, we believe that the internal thread executors employed by the Janus Runtime are capped at a certain amount of active thread jobs, and we are hitting this limit. However, we are not able to find any documentation on this scenario. It is also interesting that this seems to only happen on one of our computers, despite trying scenarios on other computers with similar specs in our team with huge numbers of events being transmitted.

I am interested to know if you have hit this issue before and how you have resolved it?

Cheers,
Matthew McNally

**Comment by gallandarakhneorg** [comment]
Dear Matthew.
We have encountered this issue. @ngaud will start working on a patch of the SARL Runtime Environment for it.
Stéphane.

**Comment by gallandarakhneorg** [comment]
Dear @hypnobunny.

I have added the "rejected task" log message in order to keep track of the tasks that are submitted and rejected.

According to the documentation of the JDK's ThreadPoolFactory, when the number of threads reaches the max pool size (by default 512 with the Janus configuration), the submitted tasks wiats into an internal queue until a thread becomes available. I believes that the tasks are rejected when the executor service is shuting down only. Nevertheless, I will check the JDK source code to ensure myself about this behavior.

Do you think some of your events is lost in your application? In other words, I would like to determine if the rejected tasks has an impact of the expected application execution.

All the best.
Stéphane.

**Comment by gallandarakhneorg** [comment]
The branch [janus_rejected_tasks](https://github.com/sarl/sarl/tree/janus_rejected_tasks) is dedicated to the solving of this issue.

**Comment by rmit-s3530405-bob-zhou** [comment]
Hi @gallandarakhneorg,

The rejected task error is causing some events to fail to be received. 
I have attached the screenshots of my output and the same in text file version. 

55 entity events were emitted to all 6 agents. Only agent A1 is printing on receipt of the events.
only 33 receipts was printed. 

Thank you very much for looking into this!

Bob


![sarl event drop 1](https://user-images.githubusercontent.com/18024518/29436379-86278e0e-83ee-11e7-8661-58b2e92e30cb.PNG)
![sarl event drop 2](https://user-images.githubusercontent.com/18024518/29436381-862abd36-83ee-11e7-9fd1-098825a8bd6f.PNG)
![sarl event drop 3](https://user-images.githubusercontent.com/18024518/29436380-8628262a-83ee-11e7-93fe-b212ffe88438.PNG)
![sarl event drop 4](https://user-images.githubusercontent.com/18024518/29436382-863bee8a-83ee-11e7-91b3-4db643f70b5d.PNG)
![sarl event drop 5](https://user-images.githubusercontent.com/18024518/29436383-863c508c-83ee-11e7-84eb-6feae67eb79e.PNG)


[sarl event.txt](https://github.com/sarl/sarl/files/1232787/sarl.event.txt)

**Comment by gallandarakhneorg** [comment]
Dear @rmit-s3530405-bob-zhou.

Could you give me the link to your code? Indeed, the way how your have written your application with  SARL may cause lost of events due the extreme parallel execution of the "on" handlers, and the calls to the capacities' functions.

I would like to be sure that the issue is in the runtime environment, not in the SARL code.

All the best.
Stéphane.

**Comment by hypnobunny** [comment]
Hi @gallandarakhneorg, thanks very much for your commitment to this issue. I am currently communicating with my team about getting you access to our code so you can better attack this issue.

Regards,
Matt

**Comment by hypnobunny** [comment]
Hi @gallandarakhneorg, @ngaud and @srodriguez, further to this thread I have sent an email with our code attached. Please let me know if you have any issues looking into it!

Regards,
Matt

**Comment by gallandarakhneorg** [comment]
Should be reproduced and fixed after Issue #751.

**Comment by gallandarakhneorg** [comment]
I confirm that this issue append in a ping pong application at IMOB with pessimistic time synchronization of events. Some events are never sent or received (we don't), and a lot of tasks are rejected.

**Comment by ssardina** [comment]
Hi @gallandarakhneorg 

Sorry I couldn't follow this up but it was in my todo list. :+1: 

You seem to have been able to reproduce the big in a simpler application? Not sure what IMOB means though.

Now, issue #751 seems to be a major one, a whole re-write. The one here is much more concrete and specific to the handling of events. Are you suggesting this issue cannot be resolved until the SRE is re-written into SARL altogether? Yikes, that is scary :-)

**Comment by gallandarakhneorg** [comment]
Dear @ssardina.
I was too busy during the past weeks to dig this issue. A couple of week ago, I have emitted the hypothesis that the current implementation of the SRE create too much `Runnable` to give to the Java executor service. The new SRE version uses a lower number of `Runnable` and it was redesigned on several points. I will check if it solve the issue, or if another executor service must be considered.

**Comment by ssardina** [comment]
Sure @gallandarakhneorg , I totally understand. I am trying to just understand how deep this issue is. But yeah, we are all so busy ... 

So there is a new SRE available?

**Comment by gallandarakhneorg** [comment]
The new SRE is not yet released. The devel version is available at: https://github.com/sarl/janus-in-sarl.

**Comment by ssardina** [comment]
Thanks. I see that is the SRE JANUS in SARL already, wow, impressive. :-)

**Comment by gallandarakhneorg** [comment]
I have decided to rewrite the SRE JANUS in SARL in order to give a small proof-of-concept that SARL grammar enables to create a "serious" application, and not only toy agent applications :-)
This version of JANUS is under review and will be released officially when the SARL team has validated it, hopefully for the version 0.7 or 0.8.

**Comment by gallandarakhneorg** [comment]
This issue is related to Issue #796.

**Comment by gallandarakhneorg** [comment]
Version 0.7 now forces the execution of the tasks even if they were rejected by the Java executor service. It means that the tasks are run within the thread that has queried to execute the rejected task.

Janus 2.0.7.0-SNAPSHOT should not log the "rejected task" error message any more.

The Janus version which was written in SARL also includes this fix and creates the Java executor service with different parameters that are expected to avoid task rejections.

I propose to close this issue, and eventually reopen it if the problem still occurs.

**Comment by ssardina** [comment]
NIce work Stephane. We have been following this issue and seems you found the root of the problem.

Now, apologies for my ignorance on Java, Runnables, etc., but what are the practical implications for the programmer regarding the fact that _"the tasks are run within the thread that has queried to execute the rejected task."_ Are thee any? Or should the programmer be careful because she/he never can now where the task will be run? Does it mean that some task can get "blocked"? Sorry again, I want to have a SARL user-programmer understanding (not a SARL developer one per se)

**Comment by gallandarakhneorg** [comment]
Running the task within the caller thread ensures that is it run and not simply skiped. But, because it is run within the caller thread, it cannot treat another task. The result could be a slow down of the application. Tasks cannot be blocked because no task contains a loop within the SARL platform code. Indeed the executor service (a java PoolThreadExecutorService) at the heart of the issue is never explicitly called by the SARL user-programmer. Tasks provided by the SARL user-programmer are executed by another executor service (a java ScheduledExecutorService) that has not the original issue.

To conclude, I think this fix will not have a strong impact on the SARL user-programmer, except a possible slow down of the application.

**Comment by e-belghache** [comment]
Hello,

I'm using SARL for a big application (real-time Big Data analysis) with a lot of agents, which communicate a lot too.

I've found **there is lost/dropped events!**

In order to check if it was because of my application (agent behaviors) or SARL (_wich is the case_), I wrote the following toy MAS composed of :

MyAgents ` 

	import io.sarl.core.Initialize
	import io.sarl.core.Logging
	import io.sarl.core.Schedules
	import io.sarl.core.DefaultContextInteractions
	
	agent MyAgent {
		uses Logging, Schedules, DefaultContextInteractions
	
		val lock = new Object
		var messages=0	
		var elapsed_time=0l
	
		on Initialize {		
			val name = 'Agent ' + occurrence.parameters.get(0) as Integer
			loggingName = name
			
			// In 10ms start sending 100 Hi to all the agents (in the defaultSpace) 
			in(10)[
				for(var i=0; i<100; i++)
					emit(new Hi(i, name))				
			]
		}	
		
		on Hi {
			synchronized(lock){
				messages++
				
	 			info('Message <'+occurrence.ID+'> received from <'+ occurrence.sender+'>. Elapsed time : ' + Math.floor((System.nanoTime-occurrence.time)/(10**6)).intValue+'ms')
				elapsed_time += System.nanoTime - occurrence.time
				
				if(messages==1000){ // 100 events x 10 agents (this included)
					info('All messages received. Mean elapsed time : ' +
							Math.floor(elapsed_time / messages / (10 ** 6)).intValue + 'ms')
					messages=0
				}		
			}
		}
	}
`
that send and receive Hi events
` 

	event Hi{
		var ID : int
		var sender : String
		val time = System.nanoTime
		
		new(ID : int, sender : String){
			this.ID=ID
			this.sender = sender
		}
	}

`
 A 'Head' agent that spawns all MyAgents `

	import io.sarl.core.Initialize
	import io.sarl.core.Schedules
	import io.sarl.core.Lifecycle
	import io.sarl.core.InnerContextAccess
	import io.sarl.core.Logging
	import io.sarl.core.DefaultContextInteractions
	
	agent Head{
		uses Schedules, Lifecycle, DefaultContextInteractions, Logging
		package var count=0
		
		on Initialize{
			loggingName = 'Head'
			info('Begin')
			
			in(10) [			
				//Spawn 10 MyAgents in the innerContext of the Head	
				for(var i=0; i<10; i++)
					spawn(MyAgent, i)			
				
				info('Done !')
			]
		}
	}
`
And finally a Launcher class of the toy MAS`

	import io.janusproject.Boot
	import io.janusproject.kernel.Kernel
	 
	class Launcher {
		static protected var janusKernel : Kernel
	
		static def main(args : String[]) : void {	
			Boot.offline = true
			janusKernel = Boot.startJanus(Head)	
		}
	}
`


Note that I use here a direct access to Janus in order to make a modification later to make a point. Nonetheless, with a Bootstrap I've got the same results/issue which is **dropped events**.  In addition, the more agents there is the bigger si the ration dropped_events/received_events ration.

I'm running my SARL applications with the following environment specs:

- SARL 0.7.0
- Janus 2.0.7.0
- Java 1.8.0_162

And machine specs:

- CPU i7-4790 @ 3.60GHz
- RAM 32Go DDR3

For this example,  in average 2200 events out of the 10 000 waited are dropped and only 4 agents receive all the events.

To remedy somehow this issue I tried using the _the Boot.setProperty_  method in _Launcher_ class to change the _janus.executors.threads.min_ property to 1 (by default it is 16), that unable the parallel executions
`

			Boot.setProperty("janus.executors.threads.min", "1000")
			janusKernel = Boot.startJanus(Head)
`
This seems working but we lose the advantage of parallelism to speed up the MAS processing and it is really important for my application.

Is there any way to fix this issue?
Thank you.

**Comment by gallandarakhneorg** [comment]
Dear @e-belghache.

According to your code, I think you are facing an initialization problem of the MAS.
The spawns of the agents takes time within `Head`. Several agents are not yet launched when an agent is emiting the events. That's why they are not receiving the events. The number of received events is then not deterministic.

For solving this issue, you have to:
1) Initialize your system by launching all agents and waiting for all the agents to be spawned (listening on `AgentSpawned` event within`Head`
2) When all agents are spawned, ask them to send the events.

**Comment by ssardina** [comment]
Good point Stephane. Now, what is in your view the "best", more clean, way
to check whether "all agents are spawned" ? In the past I have use a final
"end" event to flag that all agents have been spawn. But I was never very
happy this this solution...

**Comment by gallandarakhneorg** [comment]
Until now, I send a "ApplicationStart" event when all agents are initialized. That's not perfect for sure. Perhaps we have to start a discussion on that point. @srodriguez , what do you think about that point?

**Comment by ssardina** [comment]
OK so who sends the ApplicationStart event? Do you send it from Java or you
have like a "booting" agent?

Now that I think, how does the booting agent know that all agents have
truly finalized their creation? Say, if I create 10 agents and then emit
ApplicationStart, maybe it is done before the 9th and 10th agent have
finalized their creation (imagine a constructor that takes time)? Can this
happen?

**Comment by gallandarakhneorg** [comment]
Le 08/03/2018 à 23:49, Sebastian Sardina a écrit :
> OK so who sends the ApplicationStart event? Do you send it from Java or you
> have like a "booting" agent?

Most of the time, I create a "booting" agent. But it is possible to do from Java: the fireworks example that is embedded within the product shows how a JavaFX "start" button is used for that.

> Now that I think, how does the booting agent know that all agents have
> truly finalized their creation? Say, if I create 10 agents and then emit
> ApplicationStart, maybe it is done before the 9th and 10th agent have
> finalized their creation (imagine a constructor that takes time)? Can this
> happen?

A "AgentSpawned" event is fired after the agent is initialized (after the run of its "on Initialize", this is an example of an event that breaks the parallel execution of the event handlers while it is not finished).
So that, if your booting agent waits on AgentSpawned, you are sure that the agents are spawned.

**Comment by e-belghache** [comment]
@gallandarakhneorg  thank you very much, it works like a charm.
_(Maybe you should this should be emphasized in the tutorials and the documentation available on SARL website)_

**Comment by gallandarakhneorg** [comment]
Issue #817 concerns the enhancement of the tutorials.
